### PR TITLE
Client cache

### PR DESCRIPTION
**Pull request by dashjim**
**Issue #, if available:**
ETag and last-modified.

UT Passed.

**Description of changes:**

**Checklist**
- [x] :wave: I have run the unit tests, and all unit tests have passed.
- [ ] :warning: This pull request might incur a breaking change.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/new-image-handler/src/index.ts [hunk 1]
```diff
@@ -50,9 +52,24 @@ router.get(['/debug', '/_debug'], async (ctx) => {
 });
 
 router.get('/(.*)', async (ctx) => {
-  const { data, type } = await ossprocess(ctx, bypass);
+  const { data, type, headers } = await ossprocess(ctx, bypass);
```
Comment (wchaws) [inline source/new-image-handler/src/index.ts:55]: 这些改动都不需要了

File: source/new-image-handler/src/index.ts [hunk 2]
```diff
@@ -50,9 +52,24 @@ router.get(['/debug', '/_debug'], async (ctx) => {
...
+    let lastModified;
+    if (typeof headers.lastModified === 'string') {
+      lastModified = new Date(headers.lastModified);
+    } else {
+      lastModified = headers.lastModified;
+    }
+
+    console.log('set lastmodified: ', lastModified);
+    ctx.lastModified = lastModified;
+  }
```
Comment (wchaws) [inline source/new-image-handler/src/index.ts:72]: 这些改动都不需要了

File: source/new-image-handler/src/store.ts [hunk 3]
```diff
@@ -51,16 +52,20 @@ export class LocalStore implements IBufferStore {
...
+  public async get(p: string, beforeGetFunc?: () => void):
+  Promise<{ buffer: Buffer; type: string; headers?: IHttpHeaders}> {
     beforeGetFunc?.();
     const res = await this._s3.getObject({
       Bucket: this.bucket,
       Key: p,
     }).promise();
+
     if (Buffer.isBuffer(res.Body)) {
+      console.log(`etag: ${res.ETag}, lastmodified: ${res.LastModified}`);
```
Comment (wchaws) [inline source/new-image-handler/src/store.ts:64]: 这个 log 去掉？

File: source/new-image-handler/src/store.ts [hunk 4]
```diff
@@ -51,16 +52,20 @@ export class LocalStore implements IBufferStore {
 export class S3Store implements IBufferStore {
   private _s3: S3 = new S3({ region: config.region });
   public constructor(public readonly bucket: string) { }
-  public async get(p: string, beforeGetFunc?: () => void): Promise<{ buffer: Buffer; type: string }> {
+  public async get(p: string, beforeGetFunc?: () => void):
+  Promise<{ buffer: Buffer; type: string; headers?: IHttpHeaders}> {
```
Comment (wchaws) [inline source/new-image-handler/src/store.ts:56]: `IHttpHeaders }>` 少了个空格

File: source/new-image-handler/src/store.ts [hunk 5]
```diff
@@ -26,17 +27,17 @@ export interface IKeyValue {
   [key: string]: any;
 }
 
-export interface IBufferStore extends IStore<{ buffer: Buffer; type: string }> { };
+export interface IBufferStore extends IStore<{ buffer: Buffer; type: string; headers?: IHttpHeaders }> { };
```
Comment (wchaws) [inline source/new-image-handler/src/store.ts:30]: `headers: IHttpHeaders` 把❓问号去掉？我看你初始化都加了 headers: {}

File: source/new-image-handler/src/index.ts [hunk 6]
```diff
@@ -97,16 +114,18 @@ function getBufferStore(ctx: Koa.ParameterizedContext) {
   return DefaultBufferStore;
 }
 
-async function ossprocess(ctx: Koa.ParameterizedContext, beforeGetFn?: () => void): Promise<{ data: any; type: string }> {
+async function ossprocess(ctx: Koa.ParameterizedContext, beforeGetFn?: () => void):
+Promise<{ data: any; type: string; headers?:{etag?: string; lastModified?: string | Date } }> {
```
Comment (wchaws) [inline source/new-image-handler/src/index.ts:118]: `headers: IHttpHeaders` 改下接口类型

File: source/new-image-handler/src/index.ts [hunk 7]
```diff
@@ -97,16 +114,18 @@ function getBufferStore(ctx: Koa.ParameterizedContext) {
...
   const { uri, actions } = parseRequest(ctx.path, ctx.query);
   const bs = getBufferStore(ctx);
   if (actions.length > 1) {
     const processor = getProcessor(actions[0]);
     const context = await processor.newContext(uri, actions, bs);
+    ctx.set(context.headers);
```
Comment (wchaws) [inline source/new-image-handler/src/index.ts:124]: 这里先删了好了，挪到外面调用它的地方。

File: source/new-image-handler/src/index.ts [hunk 8]
```diff
@@ -50,9 +52,24 @@ router.get(['/debug', '/_debug'], async (ctx) => {
 });
 
 router.get('/(.*)', async (ctx) => {
-  const { data, type } = await ossprocess(ctx, bypass);
+  const { data, type, headers } = await ossprocess(ctx, bypass);
   ctx.body = data;
   ctx.type = type;
+  console.log('Set etag: ', headers?.etag);
```
Comment (wchaws) [inline source/new-image-handler/src/index.ts:58]: 加一个 `ctx.set(headers)`

**Review by wchaws — CHANGES_REQUESTED** [review]

File: source/new-image-handler/src/store.ts (hunk 5 above)
Comment (dashjim) [reply source/new-image-handler/src/store.ts:30]: LocalStore would fail, if remove that (?).

File: source/new-image-handler/src/index.ts (hunk 2 above)
Comment (dashjim) [reply source/new-image-handler/src/index.ts]: Removed.

**Review by wchaws — APPROVED** [review]

**Pull request merged**